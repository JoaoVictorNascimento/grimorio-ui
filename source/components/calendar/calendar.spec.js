import '../../../internals/test/helper';
import DayPickerRangeController from 'react-dates/lib/components/DayPickerRangeController';
import moment from 'moment';

import Calendar from './calendar-component';
import { wrap } from 'module';

/** @test {Calendar} */
describe('Calendar component', () => {
  /** @test {Calendar#render} */
  describe('#render', () => {
    let wrapper;
    const props = {
      isSingleDate: false,
      onOutsideClick: jest.fn(),
      onChange: jest.fn(),
      isOutsideRange: jest.fn(),
    };

    beforeAll(() => {
      wrapper = shallow(
        <Calendar {...props} />
      );
    });
    it('render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call changeDate', () => {
      const dates = {
        startDate: '12/03/2019',
        endDate: '15/03/2019'
      }
      const spy = jest.spyOn(wrapper.instance(), 'changeDate');
      wrapper.find(DayPickerRangeController).prop('onDatesChange')(dates);
      expect(spy).toHaveBeenCalledWith(dates);
    });

    it('should call props onChange - when focusedInput is equal endDate; and endDate is a value', () => {
      wrapper.setState({ focusedInput: 'endDate'});
      const dates = {
        startDate: '12/03/2019',
        endDate: '15/03/2019'
      }
      const spy = jest.spyOn(props, 'onChange');
      wrapper.find(DayPickerRangeController).prop('onDatesChange')(dates);
      expect(spy).toHaveBeenCalledWith({...dates, focusedInput: 'startDate'});
    });

    it('should call props onChange - when is single date', () => {
      wrapper.setProps({ isSingleDate: true });
      const dates = {
        startDate: '12/03/2019',
        endDate: '15/03/2019'
      }
      const spy = jest.spyOn(props, 'onChange');
      wrapper.find(DayPickerRangeController).prop('onDatesChange')(dates);
      expect(spy).toHaveBeenCalledWith({
        startDate: dates.startDate,
        singleDate: dates.startDate,
        focusedInput: 'startDate'
      });
    });

    it('should setState of focusedInput when focus change', () => {
      const focusedInput = 'endDate';
      wrapper.find(DayPickerRangeController).prop('onFocusChange')(focusedInput);
      expect(wrapper.state().focusedInput).toEqual('endDate');
    });

    it('on outside click - when !startDate and !singleDate', () => {
      wrapper.setState({ startDate: null, endDate: null });
      wrapper.setProps({ isSingleDate: false });
      const onOutsideClick = jest.spyOn(props, 'onOutsideClick');
      const onChange = jest.spyOn(props, 'onChange');

      wrapper.find(DayPickerRangeController).prop('onOutsideClick')();
      expect(onOutsideClick).toHaveBeenCalled();
      expect(onChange).toHaveBeenCalledWith({ startDate: null, endDate: null });
    });

    it('on outside click - when !endDate and !singleDate - set endDate as startDate', () => {
      const startDate = '12/03/2019';
      wrapper.setState({ startDate, endDate: null });
      wrapper.setProps({ isSingleDate: false });

      const onOutsideClick = jest.spyOn(props, 'onOutsideClick');
      const onChange = jest.spyOn(props, 'onChange');

      wrapper.find(DayPickerRangeController).prop('onOutsideClick')();
      expect(onOutsideClick).toHaveBeenCalled();
      expect(onChange).toHaveBeenCalledWith({ startDate, endDate: startDate });
    });

    it('on outside click - when is singleDate', () => {
      wrapper.setProps({ isSingleDate: true });

      const onOutsideClick = jest.spyOn(props, 'onOutsideClick');
      const onChange = jest.spyOn(props, 'onChange');

      wrapper.find(DayPickerRangeController).prop('onOutsideClick')();
      expect(onOutsideClick).toHaveBeenCalled();
      expect(onChange).toHaveBeenCalledWith({});
    });

    it('isOutsideRange return false - when is single date', () => {
      wrapper.setProps({ isSingleDate: true })
      const isOutsideRange = wrapper.instance().isOutsideRange(moment('30/10/2020', 'DD/MM/YYYY'));
      wrapper.setState({ focusedInput: 'endDate', startDate: moment('10/05/2019', 'DD/MM/YYYY') });

      expect(isOutsideRange).toEqual(false);
    });
    it('isOutsideRange return true - when day is after startDate and is NOT single date', () => {
      wrapper.setProps({ isSingleDate: false })
      const isOutsideRange = wrapper.instance().isOutsideRange(moment('30/10/2020', 'DD/MM/YYYY'));
      wrapper.setState({ focusedInput: 'endDate', startDate: moment('10/05/2019', 'DD/MM/YYYY') });

      expect(isOutsideRange).toEqual(true);
    });

    it('isOutsideRange should call isOutsideRange props', () => {
      const day = moment('05/01/2010', 'DD/MM/YYYY');
      wrapper.instance().isOutsideRange(day);
      wrapper.setState({ focusedInput: 'endDate', startDate: moment('10/05/2019', 'DD/MM/YYYY') });
      const spy = jest.spyOn(props, 'isOutsideRange');
      expect(spy).toHaveBeenCalledWith(day);
    })
  });
});
