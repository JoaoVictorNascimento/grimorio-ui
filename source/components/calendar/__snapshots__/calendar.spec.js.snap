// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar component #render render correctly 1`] = `
<DayPickerRangeController
  calendarInfoPosition="bottom"
  daySize={39}
  disabled={false}
  enableOutsideDays={false}
  endDate={null}
  firstDayOfWeek={null}
  focusedInput="startDate"
  hideKeyboardShortcutsPanel={true}
  horizontalMonthPadding={13}
  initialVisibleMonth={[Function]}
  isDayBlocked={[Function]}
  isDayHighlighted={[Function]}
  isFocused={false}
  isOutsideRange={[Function]}
  isRTL={false}
  keepOpenOnDateSelect={false}
  minimumNights={0}
  monthFormat="MMMM YYYY"
  navNext={null}
  navPrev={null}
  noBorder={false}
  noNavButtons={false}
  numberOfMonths={2}
  onBlur={[Function]}
  onClose={[Function]}
  onDatesChange={[Function]}
  onFocusChange={[Function]}
  onNextMonthClick={[Function]}
  onOutsideClick={[Function]}
  onPrevMonthClick={[Function]}
  onShiftTab={[Function]}
  onTab={[Function]}
  orientation="horizontal"
  phrases={
    Object {
      "calendarLabel": "Calendar",
      "chooseAvailableDate": [Function],
      "chooseAvailableEndDate": [Function],
      "chooseAvailableStartDate": [Function],
      "dateIsSelected": [Function],
      "dateIsUnavailable": [Function],
      "enterKey": "Enter key",
      "escape": "Escape key",
      "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
      "homeEnd": "Home and end keys",
      "jumpToNextMonth": "Move forward to switch to the next month.",
      "jumpToPrevMonth": "Move backward to switch to the previous month.",
      "keyboardShortcuts": "Keyboard Shortcuts",
      "leftArrowRightArrow": "Right and left arrow keys",
      "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
      "moveFocusByOneMonth": "Switch months.",
      "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
      "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
      "openThisPanel": "Open this panel.",
      "pageUpPageDown": "page up and page down keys",
      "questionMark": "Question mark",
      "returnFocusToInput": "Return to the date input field.",
      "selectFocusedDate": "Select the date in focus.",
      "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
      "upArrowDownArrow": "up and down arrow keys",
    }
  }
  renderCalendarInfo={null}
  renderDayContents={null}
  renderMonthElement={null}
  renderMonthText={null}
  showKeyboardShortcuts={false}
  startDate={null}
  verticalHeight={null}
  weekDayFormat="dd"
  withPortal={false}
/>
`;
